dask/tests/test_local.py

```diff
@@ -170,3 +172,22 @@ def append(i):
     get_sync(dsk, "y")
 
     assert L == sorted(L)
+
+
+def test_complex_ordering():
+    da = pytest.importorskip("dask.array")
+    from dask.diagnostics import Callback
+
+    actual_order = []
+
+    def track_order(key, dask, state):
+        actual_order.append(key)
+
+    x = da.random.normal(size=(20, 20), chunks=(-1, -1))
+    res = (x.dot(x.T) - x.mean(axis=0)).std()
+    dsk = dict(res.__dask_graph__())
+    exp_order_dict = order(dsk)
+    exp_order = sorted(exp_order_dict.keys(), key=exp_order_dict.get)
+    with Callback(pretask=track_order):
+        get_sync(dsk, exp_order[-1])
+    assert actual_order == exp_order
```

Comment:
@mrocklin So this is the first draft of a test like you talked about (I think). However, it only uses one large chunk so even though the ordering is correct I'm not sure it is actually testing what you were hoping. If I make the array more than one chunk then the order does not match. `dask.order` expects all the `normal` and `transpose` tasks to be done for an entire row of chunks (I think) before it does the first `mean_chunk`.

The pre-concurrent.futures scheduler matches the current scheduler but does not match `dask.order` exactly. The scheduler prefers to work on a chunk as far as it can which apparently means doing `normal` -> `transpose` -> `mean_chunk` and then going on to the next chunk. This is because it immediately processes any tasks that were recently added to the "ready" list (dependencies have all been processed and cached).

Reply:
I looked at the local scheduler more and the only way to resolve this that I can think of is to only add `waiting` tasks to `ready` when the task before it (order-wise) is in the `ready`/`running`/`finished` list already. However, this part of the code doesn't know anything about the `dask.order` order, it is only using the pre-sorted lists of tasks. It says "ah this waiting task has no waiting dependencies, then I should run it". Eager execution I guess.

Reply:
cc @eriknw should we trust order or should we trust runtime decisions like this?

Reply:
I'll need to take a closer look to understand the current behavior to provide a more informed opinion.

Generally, I would expect the local scheduler to trust and defer to `dask.order` within reason unless it has a good reason not to.  A good reason not to would be that a dynamic scheduler can know the sizes of data, and can prioritize releasing large data.  A dynamic scheduler may also estimate runtime and data sizes, and is better able to react to the current state of execution.

Didn't the previous threaded scheduler basically run the best available task according to `dask.order`?  If there is now under-informed greedy, LIFO, or FIFO behavior occurring, I do worry about some workloads behaving poorly.  `dask.order` is complicated for good reasons!  The local scheduler should probably be stress-tested similar to how I stress-test `dask.order`.  I may be able to get some time to do this.

Reply:
> Didn't the previous threaded scheduler basically run the best available task according to dask.order? If there is now under-informed greedy, LIFO, or FIFO behavior occurring, I do worry about some workloads behaving poorly. 

The main thing in this PR is that it is restoring pre-2021.04.0 behavior. This PR basically takes away the "oh I think we can do this better" decisions that were made in the concurrent.futures PR which maybe tried to do too much for a single PR (concurrent.futures, batch task submission, reordering tasks to avoid reverse sorting).

The threaded scheduler will keep processing down (up?) a branch of the tree until it can't any longer. This is against what `dask.order` would prefer where `dask.order` would rather stop processing a certain chunk and balance out the processing of other chunks. Which dependencies get processed first are influenced by `dask.order` so overall order of execution for branches further up the tree (not directly depending on leaves) should remain intact.

* Could the threaded scheduler "listen" to `dask.order` more? Absolutely.
* Was it listening more before the concurrent.futures switch? No. This PR has equivalent behavior.
* Was it listening more after the concurrent.futures switch? Even less so. It was directly contradicting what `dask.order` was trying to do.